### ISSUE TITLE
Perhaps `SearchResult.download()` should accept an author "priority" order

### ISSUE THREAD

**Issue by christinahedges** [lightkurve/lightkurve#1048]
Currently in lightkurve we have this awesome feature of being able to get light curves from different authors:

![image](https://user-images.githubusercontent.com/14965634/116444441-943b0500-a809-11eb-94dd-4b9063890015.png)

This is great! Sometimes I want to get a lot of light curves, on lots of different targets, and I have a preference for which author I want the light curves from. E.g. I might want `TESS-SPOC` wherever I can, and then if that is unavailable I'd like `SPOC`, and then if that is unavailable I'd like `QLP`.

Right now, if I `download_all`, I will get 6 files (i.e. three versions of each of the two sectors, from the three authors.)

It would be possible to have something like:
`sr.download_all(author_priority=['TESS-SPOC', 'SPOC', 'QLP'])`

Which would potentially download just two files from `TESS-SPOC`

If I had this search result:
![image](https://user-images.githubusercontent.com/14965634/116444769-eed46100-a809-11eb-9168-4215f01287c1.png)

And used this command:
`sr.download_all(author_priority=['TESS-SPOC', 'SPOC', 'QLP'])`

It would just download one `QLP` light curve, because that's what's available.

The benefit of this is that it would enable people who want to step through a catalog of sources to get the "best" light curve for their science case. For example, people who want to make streamline pipelines for exoplanet discovery.

I'm opening this up as a discussion because I want to see what others think.

**Comment by orionlee** [comment]
I found myself mentally doing something similar in a notebook environment, typically in browsing and viewing HLSPs from older sectors.

Having said that, I don't know if it'll be one of those nifty features that is actually useful; or one of those that sounds good on paper, but doesn't quite work well practically.

In the meantime, I have created a helper function, [`filter_by_priority()`](https://gist.github.com/orionlee/0585e67dc245d4b95bd948a8aa284b64), that does basically the same thing. We could possibly play with it, and see how it helps or does not help in practice. 

Example usage:
```python
    filter_by_priority(sr, author_priority=['TESS-SPOC', 'SPOC', 'QLP'].download_all()
```

Fine points:
- A complication I notice is that we probably also need to specify a priority for `exptime`: for starndard `SPOC` products, there are cases that both 2 minute cadence and 20 second cadence products are available in the same sector.

**Comment by orionlee** [comment]
I have updated the `filter_by_priority` helper to handle both author and exptime. Ping me if there is there is interests in incorporating it into `lightkurve`

Example usage:
```
filter_by_priority(sr,
                   author_priority=['SPOC', 'TESS-SPOC', 'QLP'],
                   exptime_priority=['short', 'long', 'fast']
                   ).download_all()
```

The helper:
https://gist.github.com/orionlee/0585e67dc245d4b95bd948a8aa284b64

**Comment by barentsen** [comment]
Thank you for raising this idea @christinahedges, and for posting an example solution @orionlee!

For completeness, I'll add that an author priority system can of course also be implemented using multiple queries, e.g.:

```python
author_priority = ["SPOC", "TESS-SPOC", "QLP"]
lc = None
for author in author_priority:
    result = lk.search_lightcurve("Pi Men", sector=1, author="SPOC")
    if len(result):
        lc = result.download()
        continue
```

While multiple queries offer a fairly simple solution, @christinahedges did point out that this would incur a speed penalty.  Adding an `author_priority` (and `exptime_priority`?) feature would add quite a bit of complexity to the search module however, so we'd want to be very certain that the trade-off between complexity and speed-up is worth it.

For the time being, I think I'd like to see a few more use cases and solutions discussed on this thread before we consider adding this as a new feature.

**Comment by orionlee** [comment]
I also feel we should sit with it a while to see if it's worth supporting it.

I feel the complexity is not necessarily just in the codes, but also the API we expose. Explaining the feature seems to be quite difficult, until the users run across cases that they wish something like it.

Here is what I'd propose if we do decide to support it. Add a new method, `unique_missions()`, rather than introducing it to I'll use an example.

```python
sr = lk.search("tic180695581", mission="TESS")
sr = sr.unique_missions(priority=dict(author=["SPOC", "QLP", "TESS-SPOC"], exptime=["short", "long", fast"]))
lcc = sr.download_all()
```

I feel that having the feature exposed as its own function is better than adding it to `search` or `download` because:
- from api documentation perspective: users who don't care about it won't be burdened when they need to read the doc on `search` / `download`.
- having is own functions also make its purpose clearer when one reads codes.

**Comment by barentsen** [comment]
> I feel the complexity is not necessarily just in the codes, but also the API we expose.

I fully agree!  In fact I think reducing API complexity is 100x more important than reducing implementation complexity.

> I feel that having the feature exposed as its own function is better

Yes, this is a good idea!  I agree that this feature would best be added as a `SearchResult` method as opposed to adding extra parameters to `search()` or `download()`.
